src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -1100,37 +1132,227 @@ private void PopulateParamsFromRestRequest(Dictionary<string, object?> parameter
                         );
             }
 
-            if (!isAuthorized)
+            return isAuthorized;
+        }
+
+        /// <summary>
+        /// Performs authorization checks on entity level permissions and field level permissions for every entity and field
+        /// referenced in a GraphQL mutation for the given client role.
+        /// </summary>
+        /// <param name="context">Middleware context.</param>
+        /// <param name="clientRole">Client role header value extracted from the middleware context of the mutation</param>
+        /// <param name="entityName">entity name</param>
+        /// <param name="entityName">Top level entity name.</param>
+        /// <param name="operation">mutation operation</param>
+        /// <param name="inputArgumentName">Name of the input argument (differs based on point/multiple mutation).</param>
+        /// <param name="parametersDictionary">Dictionary of key/value pairs for the argument name/value.</param>
+        /// <exception cref="DataApiBuilderException">Throws exception when an authorization check fails.</exception>
+        private void AuthorizeEntityAndFieldsForMutation(
+            IMiddlewareContext context,
+            string clientRole,
+            string entityName,
+            EntityActionOperation operation,
+            string inputArgumentName,
+            IDictionary<string, object?> parametersDictionary
+            )
+        {
+            IInputField schemaForArgument = context.Selection.Field.Arguments[inputArgumentName];
+
+            // Dictionary to store a mapping from entityName to all field names being referenced from that entity in the mutation.
+            Dictionary<string, HashSet<string>> entityAndFieldsToAuthorize = new();
+            object? parameters;
+            if (parametersDictionary.TryGetValue(inputArgumentName, out parameters))
+            {
+                // Get all the entity names and field names referenced in the mutation.
+                PopulateMutationFieldsToAuthorize(entityAndFieldsToAuthorize, schemaForArgument, entityName, context, parameters, _runtimeConfigProvider.GetConfig());
+            }
+
+            // List of entities which are being referenced in the mutation.
+            IEnumerable<string> entityNames = entityAndFieldsToAuthorize.Keys;
+
+            // Perform authorization check at entity level.
+            foreach(string entityNameInMutation in entityNames)
             {
-                throw new DataApiBuilderException(
-                    message: "Unauthorized due to one or more fields in this mutation.",
-                    statusCode: HttpStatusCode.Forbidden,
-                    subStatusCode: DataApiBuilderException.SubStatusCodes.AuthorizationCheckFailed
-                );
+                if (!_authorizationResolver.AreRoleAndOperationDefinedForEntity(entityNameInMutation, clientRole, operation))
```

Comment:
Why do we need to separately perform authZ at entity level if we are anyway doing the same (albeit in more detail) at the field level? If the role and operation is not defined for entity, the field level authZ we perform below should also return that the field is unauthorized. In essence, arent the checks we do in AreFieldAuthorizedForEntity a superset of `AreRoleAndOperationDefinedForEntity`?

Reply:
Right, I just looked at the implementation of the AuthorizationResolver.AreColumnsAllowedForOperation(). It has checks for whether the role and action is defined for an entity. So, yes we don't need the `AreRoleAndOperationDefinedForEntity` check.